src/Language/PureScript/CoreFn/Laziness.hs

```diff
@@ -0,0 +1,504 @@
+-- declarations as the following sum type:
+
+data RecursiveGroupItem e = EagerBinding Ann e | LazyDefinition e | LazyBinding Ann
+  deriving Functor
+
+-- |
+-- Transform a recursive binding group, reordering the bindings within when a
+-- correct initialization order can be statically determined, and rewriting
+-- bindings and references to be lazy otherwise.
+--
+applyLazinessTransform :: ModuleName -> [((Ann, Ident), Expr Ann)] -> ([((Ann, Ident), Expr Ann)], Any)
+applyLazinessTransform mn rawItems = let
+
+  -- Establish the mapping from names to ints.
+  rawItemsByName = M.fromList $ (snd . fst &&& first fst) <$> rawItems
+  maxIdx = M.size rawItemsByName - 1
+  rawItemsByIndex = A.listArray (0, maxIdx) $ M.elems rawItemsByName
+  names = M.keysSet rawItemsByName
+
+  -- Now do the first delay/force traversal of all the bindings to find
+  -- references to other names in this binding group.
+  --
+  -- The parts of this type mean:
+  -- D is the maximum force (or Nothing if unknown) with which the identifier C
+  -- is referenced in any delay-B position inside the expression A.
+  --
+  -- where A, B, C, and D are as below:
+  --                A           B (keys)           C (keys)           D
+  findReferences :: Expr Ann -> IM.MonoidalIntMap (IM.MonoidalIntMap (Ap Maybe (Max Int)))
+  findReferences = (getConst .) . onVarsWithDelayAndForce $ \delay force _ -> \case
+    Qualified mn' ident | all (== mn) mn', Just i <- ident `S.lookupIndex` names
+      -> Const . IM.singleton delay . IM.singleton i $ coerceForce force
+    _ -> Const IM.empty
+
+  -- The parts of this type mean:
+  -- D is the maximum force (or Nothing if unknown) with which the identifier C
+  -- is referenced in any delay-B position inside the binding of identifier A.
+  --
+  -- where A, B, C, and D are as below:
+  --                     A    B (keys)           C (keys)           D
+  refsByIndex :: A.Array Int (IM.MonoidalIntMap (IM.MonoidalIntMap (Ap Maybe (Max Int))))
+  refsByIndex = findReferences . snd <$> rawItemsByIndex
+
+  -- Using the approach explained above, traverse the reference graph generated
+  -- by `refsByIndex` and find all reachable names.
+  --
+  -- The parts of this type mean:
+  -- D is the maximum force with which the identifier C is referenced,
+  -- directly or indirectly, during the initialization of identifier A. B is
+  -- Nothing if the analysis of A was inconclusive and A might need the entire
+  -- binding group.
+  -- 
+  -- where A, B, C, and D are as below:
+  --                           A    B      C (keys)           D
+  reachablesByIndex :: A.Array Int (Maybe (IM.MonoidalIntMap (Max Int)))
+  reachablesByIndex = searchReachable maxIdx $ \(i, force) ->
+    getAp . flip IM.foldMapWithKey (dropKeysAbove force $ refsByIndex A.! i) $ \delay ->
+      IM.foldMapWithKey $ \i' force' ->
+        Ap $ IM.singleton i' . Max . (force - delay +) <$> uncoerceForce force'
```

Comment:
This function calculates `h = f - d + g`, but also accounts for expressions where an identifier occurs multiple times with different force. For example:
```
-- given `C` is some identifier
C = ...
-- and `C` is used multiple times in the initialization
-- of `A` with forces, `x`, `y`, and `z`...
A = C#_!x ... C#_!y ... C#_!z
-- Then `D` is the maximum of those force values
-- (e.g. `max [x, y, z]`)
```
Or put differently, who cares if `C` is called with force 1 or force 2 because if `C` is called with force `5`, that's the case we need to care about.

Reply:
Correct, but note that this covers not just the places that `C` appears in the initializer of `A` as you illustrated, but also all the other places that `C` occurs in the binding group that could be indirectly used during the evaluation of the initializer of `A`.